arci-ros2/src/navigation.rs

```diff
@@ -63,59 +63,90 @@ impl Navigation for Ros2Navigation {
+                                },
+                                pose: to_pose_msg(&goal),
+                            },
+                            ..Default::default()
+                        };
 
-        let has_reached_set = has_reached.clone();
-        let result_cb = Box::new(move |r: NavigateToPose::Result| {
-            info!("final result {:?}", r);
-            *has_reached_set.lock().unwrap() = Some(r);
+                        is_available.await.unwrap();
+                        let send_goal_request = action_client.send_goal_request(goal).unwrap();
+                        let (goal, result, feedback) = send_goal_request.await.unwrap();
+                        *current_goal_clone.lock().unwrap() = Some(goal.clone());
+                        tokio::spawn(
+                            async move { feedback.for_each(|_| std::future::ready(())).await },
+                        );
+                        result.await.unwrap(); // TODO: handle goal state
+                        is_done.store(true, Ordering::Relaxed);
+                    });
+                    loop {
+                        node.lock()
+                            .unwrap()
+                            .spin_once(std::time::Duration::from_millis(100));
+                        tokio::task::yield_now().await;
+                        if is_done_clone.load(Ordering::Relaxed) {
+                            break;
+                        }
+                    }
+                    *current_goal.lock().unwrap() = None;
+                    let _ = sender.send(());
+                });
         });
-        self.action_client
-            .send_goal_request(goal, cb, feedback_cb, result_cb)
-            .unwrap();
-        let spin_node = self.node.clone();
-        let start_time = std::time::Instant::now();
         let wait = WaitFuture::new(async move {
-            const SLEEP_DURATION: std::time::Duration = std::time::Duration::from_micros(100);
-            while has_reached.lock().unwrap().is_none() && start_time.elapsed() < timeout {
-                spin_node.lock().unwrap().spin_once(SLEEP_DURATION);
-                std::thread::sleep(SLEEP_DURATION);
-            }
-            // TODO: Handle the result and timeout
+            let _ = receiver.await;
             Ok(())
         });
 
         Ok(wait)
     }
 
     fn cancel(&self) -> Result<(), Error> {
-        todo!();
+        // TODO: current_goal is None until send_goal_request.await is complete.
+        //       Therefore, if cancel is called during that period, it will not work correctly.
+        if let Some(current_goal) = self.current_goal.lock().unwrap().take() {
+            std::thread::spawn(move || {
+                tokio::runtime::Builder::new_current_thread()
+                    .enable_all()
+                    .build()
+                    .unwrap()
+                    .block_on(async move {
+                        let _ = current_goal.cancel().unwrap().await;
```

Comment:
It might be better to return some error if cancel() failed instead of unwrap() here.

Reply:
done